rleplus/internal/bitvector.go

```diff
@@ -0,0 +1,187 @@
+package bitvector
+
+import "errors"
+
+var (
+	// ErrOutOfRange - the index passed is out of range for the BitVector
+	ErrOutOfRange = errors.New("index out of range")
+)
+
+// BitNumbering indicates the ordering of bits, either
+// least-significant bit in position 0, or most-significant bit
+// in position 0.
+//
+// It it used in 3 ways with BitVector:
+// 1. Ordering of bits within the Buf []byte structure
+// 2. What order to add bits when using Extend()
+// 3. What order to read bits when using Take()
+//
+// https://en.wikipedia.org/wiki/Bit_numbering
+type BitNumbering int
+
+const (
+	// LSB0 - bit ordering starts with the low-order bit
+	LSB0 BitNumbering = iota
+
+	// MSB0 - bit ordering starts with the high-order bit
+	MSB0
+)
+
+// BitVector is used to manipulate ordered collections of bits
+type BitVector struct {
+	BytePacking BitNumbering
+	Buf         []byte
+	Len         int
+}
+
+// NewBitVector constructs a new BitVector from a slice of bytes.
+//
+// The bytePacking parameter is required to know how to interpret the
+// bit ordering within the bytes.
+func NewBitVector(buf []byte, bytePacking BitNumbering) *BitVector {
+	return &BitVector{
+		BytePacking: bytePacking,
+		Buf:         buf,
+		Len:         len(buf) * 8,
+	}
+}
+
+// Push adds a single bit to the BitVector.
+//
+// Although it takes a byte, only the low-order
+// bit is used, so just use 0 or 1.
+func (v *BitVector) Push(val byte) {
+	if v.Len%8 == 0 {
+		v.Buf = append(v.Buf, 0)
+	}
+
+	switch v.BytePacking {
+	case LSB0:
+		v.pushLSB0(val)
+	default:
+		v.pushMSB0(val)
+	}
+
+	v.Len++
+}
+
+// Get returns byte either 0, 1
+func (v *BitVector) Get(idx int) (byte, error) {
+	if idx >= v.Len {
+		return 0, ErrOutOfRange
+	}
+
+	switch v.BytePacking {
+	case LSB0:
+		return v.getLSB0(idx), nil
+	default:
+		return v.getMSB0(idx), nil
+	}
+}
+
+// Extend adds up to 8 bits to the receiver, in the order indicated
+//
+// Given a byte b == 0b11010101
+// v.Extend(b, 4, LSB0) would add <1, 0, 1, 0>
+// v.Extend(b, 4, MSB0) would add <1, 1, 0, 1>
+func (v *BitVector) Extend(val byte, count uint, order BitNumbering) {
+	if count > 8 {
+		count = 8
+	}
+
+	switch order {
+	case LSB0:
+		v.extendLSB0(val, count)
+	default:
+		v.extendMSB0(val, count)
+	}
+}
+
+// Take reads up to 8 bits at the given index, and returns them in a byte.
+// The byte packing is determined by the order parameter.
+//
+// Given a BitVector < 1, 1, 0, 1, 0, 1, 0, 1 >
+// v.Take(1, 4, LSB0) would return 0b00001011
+// v.Take(1, 4, MSB0) would return 0b00001101
```

Comment:
There is something weird about this method's interpretation of `MSB0`.  To preserve symmetry across the package `v.Take(1, 4, MSB0)` would be `0b11010000`.  If the output of `Take` in `MSB0` mode is actually interpreted as an `MSB0` byte then it is multiplied by something like `2^(8-count)` when compared with the `LSB0` output of Take.

I haven't gotten to consumption of `Take` and I'm sure it makes sense when doing RLE+, but the difference and or motivation is probably worth explaining in the comment here.

Reply:
👍 Yeah, the semantics of `Take` and `Extend` are confusing.  Will rationalize.